File: LeetCode75_Artur/Linked List/2095_Delete_the_Middle_Node_of_a_Linked_List.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode* deleteMiddle(ListNode* head)
    {
        if(!head->next)
            return nullptr;

        ListNode *slow = head;
        ListNode *fast = head->next;

        while(true)
        {
            if(!fast->next || !fast->next->next)
            {
                ListNode *next = slow->next->next;
                slow->next->next = nullptr;
                slow->next = next;
                break;
            }
            slow = slow->next;
            fast = fast->next->next;
        }

        return head;
    }
};

File: LeetCode75_Artur/Linked List/2095_Delete_the_Middle_Node_of_a_Linked_List.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def deleteMiddle(self, head: ListNode):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if not head.next:
            return None
        
        slow = head
        fast = head.next
        
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        deletedNode = slow.next
        slow.next = deletedNode.next
        deletedNode.next = None
        return head
